Add new 'group_execute' resource

cottsay opened the pull request:
The Chef 'execute' resource allows you to specify user and group permissions with which to execute the command. Unfortunately, this doesn't take into account seccondary groups listed in /etc/group.

We're currently using those seccondary groups to enable a user (process) to access the GPG vault, and the reprepro commands invoked by chef will need to perform some signing, and will therefore need permission.

The new resource defined by this change is nearly identical to the 'execute' resource, but adds an additional 'seccondary_groups' property. The resource then augments the supplied command by prefixing it with calls to 'sg <group-name>' to chain through the given groups. If a 'group' value was specified by the resource, that group is added to the call chain last to ensure it is the primary group after the other groups have been added.

Some examples, assuming that `jenkins-agent` is a member of the `gpg-vault` group:

A normal `execute` invocation:
```
execute 'id' do
  user 'jenkins-agent'
  group 'jenkins-agent'
end
```
Result: `uid=1001(jenkins-agent) gid=1001(jenkins-agent) groups=1001(jenkins-agent),0(root)`

An invocation using `group_execute`:
```
group_execute 'id' do
  user 'jenkins-agent'
  group 'jenkins-agent'
  secondary_groups ['gpg-vault']
end
```
Result: `uid=1001(jenkins-agent) gid=1001(jenkins-agent) groups=1001(jenkins-agent),0(root),997(gpg-vault)`
The actual command invoked is: `sg gpg-vault sg\ jenkins-agent\ id`

cottsay committed:
Add new 'group_execute' resource

The Chef 'execute' resource allows you to specify user and group
permissions with which to execute the command. Unfortunately, this
doesn't take into account seccondary groups listed in /etc/group.

We're currently using those seccondary groups to enable a user (process)
to access the GPG vault, and the reprepro commands invoked by chef will
need to perform some signing, and will therefore need permission.

The new resource defined by this change is nearly identical to the
'execute' resource, but adds an additional 'seccondary_groups' property.
The resource then augments the supplied command by prefixing it with
calls to 'sg <group-name>' to chain through the given groups. If a
'group' value was specified by the resource, that group is added to the
call chain last to ensure it is the primary group after the other groups
have been added.

nuclearsandwich commented on libraries/group_execute_provider.rb:
```diff
@@ -0,0 +1,28 @@
+        if new_resource.secondary_groups
+          groups = new_resource.secondary_groups
+          if group
+            # Always do the 'primary' group last in the call chain
+            groups = [group] + groups
```
Unless I'm misreading, it seems like the comment and statement disagree.
The comment says the primary group should be last but then concatenates an array containing just the primary group to the head of the array of groups.

cottsay replied:
The call stack is created from the inside out, which makes the shell escaping easier. So the last group in the list is the first one to get evaluated as it works in towards the actual command. The order of the groups seems to have no bearing on the behavior, beyond the fact that the last (inner-most) call ends up setting the primary group.

cottsay replied:
If it would make it more clear, we could make this `groups + [group]` as long as we reverse the iteration order on line 19.

cottsay replied:
I went ahead and reversed the iteration in 702abf3.

nuclearsandwich commented on libraries/group_execute_provider.rb:
```diff
@@ -0,0 +1,28 @@
+            groups = [group] + groups
+          end
+          groups.each do |secondary_group|
+            command = "sg #{secondary_group} #{Shellwords.shellescape(command)}"
```
Does this work as expected if the original command is an Array rather than a String?

cottsay replied:
Oh, I remember seeing that was a possible type for the argument and then blocking it from my mind. No, it's not handled. Would it be OK if I just asserted it as unsupported? I don't really want to deal with escaping spaces in arguments prior to joining the array into a flat string.

nuclearsandwich replied:
Yeah I think as long as it blows up rather than silently does something unexpected that's fine.

In vanilla Ruby I'd just raise an ArgumentError but I'm not sure if there is a more chefly way to re-declare an inherited parameter. If there's nothing obvious in the docs you've already encountered I'd suggest just dumping the following at the start of the command method.
```ruby
raise ArgumentError, "Only String commands are supported" unless command.is_a? String
```

cottsay replied:
I think I might know a way, let me give it a shot. Otherwise, I'll just raise as you suggested.

cottsay replied:
Alright, after b73f644, you'll now be greeted with this message:
```
 Chef::Exceptions::ValidationFailed
       ----------------------------------
       Option command must be a kind of [String]!  You passed ["echo", "foo"].
```

nuclearsandwich replied:
Great! I've never overloaded/subclassed a chef resource like this so you're blazing trails I have not.

cottsay committed:
Override 'command' for the group_execute resource

This way we can assert that the 'command' value's type is a string
instead of an array, which is also supported by the 'execute' resource
but is not supported by 'group_execute'.

cottsay committed:
Iterate over the group list backwards

Since the command is built from the inside-out to make the shell
escaping simpler, we want to start with the last group in the list,
which will then be switched to last in the call stack.

This should result in no functional change in behavior, but might make
the code a little bit easier to understand.

nuclearsandwich commented on libraries/group_execute_provider.rb:
```diff
@@ -0,0 +1,28 @@
+          groups = new_resource.secondary_groups
+          if group
+            # Always do the 'primary' group last in the call chain
+            groups = groups + [group]
```
```suggestion
            groups << group
```

This is mostly a style suggestion. `<<` is an append operator and aside from being more common it avoids an array allocation for `[group]`.

nuclearsandwich reviewed (approved):
A non-blocking style suggestion is my only remaining comment.

If this ends up bearing a lot of load for us we should consider trying to get secondary groups supported in the upstream execute resource.

cottsay committed:
Update libraries/group_execute_provider.rb

Co-authored-by: Steven! Ragnarök <nuclearsandwich@users.noreply.github.com>
